FSE: Add script to automatically sync changes to your wpcom sandbox

noahtallen opened the pull request:
### Changes proposed in this Pull Request
* Adds a script to rsync FSE to your sandbox when there are changes
* Allows you to pass `--sync` as an arg to the dev and build commands in order to sync to your sandbox

### SSH config
A prerequisite for this script is being able to reference a named SSH config to sync changes to your sandbox. 

`vi ~/.ssh/config` reveals this entry for me with some details renamed:

```
# If you don't add your keys to the SSH agent, it'll continuously ask you for your password!
Host *
    AddKeysToAgent yes
    IdentityFile ~/.ssh/id_rsa

Host wpcom-sandbox
    User yoursandboxusernamehere
    HostName yoursandboxurlhere.word press.com
```

I figure it's likely that many of us already have this set up. All that you'll have to do is add an entry for `wpcom-sandbox` pointing at your sandbox, and then the script should work without any further setup. Way better than all the pain our new devs have had to go through getting unison set up!

### Test it out 
First, make sure you are **not** running any existing unison or build processes.

1. Make sure you have the above SSH config file set up.
2. `svn up` on your sandbox to get the latest changes. Also note that rsync should work more quickly if you have an active SSH session to your sandbox running at the same time
2. Run `nvm use && npm i` in the root of wp-calypso to get everything up to date.
3. Run `npx lerna run dev --scope='@automattic/full-site-editing' --stream -- -- --sync`

Obviously, the above command is pretty nasty because you have to pipe the args through npx, lerna, npm, and into a node script. I would suggest making an alias for this command in your `~/.bash_profile`.

Once this is running, you should see something like `Running the following commands: dev:*,wpcom-sync` near the bottom of the output.

To test that it's working:
1. Make a modification to a PHP file like `full-site-editing-plugin.php`. Then in your sandbox, open up `wp-content/plugins/full-site-editing-plugin/dev/full-site-editing-plugin.php` and make sure you see the change reflected.
2. Make a modification to a JS file (something you can see in the UI). Then, with a site and the API sandboxed, visit the FSE editor for that site. Make sure you see that change in the UI. (Also verify in the console that you are loading the `dev` version.)

### To do:
- [ ] Display output of sync and build. Currently it seems to be hidden.

noahtallen committed:
Add FSE npm script which rsyncs files to wpcom-sandbox SSH entity

noahtallen committed:
Move script to bash for ease of use

noahtallen committed:
Try to run a dev and build script separately

noahtallen committed:
Node script to conditionally run sync with build/dev

noahtallen committed:
Fix log output

Addison-Stavlo commented:
I just tested this out.  

The issue I am having is that once I have the script running, it requires me to type in my passphrase every time I save a change and want it sync'd.

The changes are reflected in the files, but I had to type in my passphrase multiple times.  Once to do the initial sync, a second time to sync the first change to test, and a third time to sync the second change. etc.

noahtallen commented:
@Addison-Stavlo Ah, that makes sense. I think you would have to save your SSH password in the keychain or something like that. In my ssh config file I also have:

```
Host *
  AddKeysToAgent yes
  UseKeychain yes
  IdentityFile ~/.ssh/id_rsa
```

Addison-Stavlo commented:
Ah I see.  I was not adding it to my keychain previously because of security practices/concerns (but I'm not sure if this is actually a big issue to do it this way?).  

I will note that using the unison sync that I only had to type in my passphrase once on initially starting it and all subsequent save/syncs did not need it (without having it keychained).

noahtallen commented:
Gotcha. I think that makes sense -- `rsync` is a one off thing, so it isn't watching for changes like unison did. We basically run the rsync command every time we have new changes, so it asks you for the password every time it runs the command. As far as i can tell there's no way to get around this unless we program the script to ask for your password upfront and then include it in the SSH request every time. That might actually be a good idea

Addison-Stavlo reviewed (approved):
Nice, after updating the ssh config file to use that this works well!

Getting started with developing in the fse plugin will be way easier now without having to set up the unison file and without having to worry about updating to the correct version number each time!

noahtallen committed:
display output in build script

noahtallen commented:
Updated to include build output :)

You can get the following output with this command: `npx lerna run dev --scope='@automattic/full-site-editing' --stream --no-prefix -- -- --sync`

<img width="1366" alt="Screen Shot 2020-01-20 at 1 02 52 PM" src="https://user-images.githubusercontent.com/6265975/72757807-d47ef280-3b85-11ea-9b64-aff4adeeb778.png">

Here is what each number indicates:
1. There was an error in the webpack build
2. The script which syncs to wpcom detected new changes in these files
3. Webpack successfully completed the build for the module listed on the left (`dev:full site editing`)
4. This is a normal PHP file (untouched by the build step) which I modified. This change was detected and synced to wpcom.
